### ISSUE TITLE
Npc dialog window is vanishing

### ISSUE THREAD

**Issue by lelouch22** [rathena/rathena#2177]
* **rAthena Hash**: 3e625df

* **Client Date**: 2014-10-22bRagexe

* **Server Mode**: Pre-Renewal

* **Description of Issue**: 
  * Result: There is a chance that npc dialog window will got vanished after you clicked an npc (ex. Quest Shop) and you got stock which you cannot talk to any npc unless you reconnect to the game to talk to that npc again.
  * Expected Result: This shouldn't occur even you stay in that npc with npc dialog window it will not got vanished.

**Comment by anacondaqq** [comment]
Can confirm. Later will provide steps how reproduce.
Modifications not really matters for the problem, exe too. Issue can happen on 2012(?) / 2013 / 2014 / 2015 exes. I though it's custom mods at mine emulator, but after testing on clean rathena i can confirm that the issue exist too in rAthena and not really related to game exe. Server mode any: pre-re or re, other settings are default, but tried different scenarios and always issue happened. 

After debugging some time ago, the issue were related for calling:
`clif_scriptclose()` and select / or NPC dialogue removed.

**Comment by Lemongrass3110** [comment]
This sounds just like the secure npc timeout? Or are you talking about something else?

**Comment by lelouch22** [comment]
Yea i think its connected to the secure npc timeout but mine is already disabled and its still occuring.

```
/**
 * Optional NPC Dialog Timer
 * When enabled all npcs dialog will 'timeout' if user is on idle for longer than the amount of seconds allowed
 * - On 'timeout' the npc dialog window changes its next/menu to a 'close' button
 * Comment to disable the timer.
 **/
//#define SECURE_NPCTIMEOUT
```

**Comment by Jeybla** [comment]
> For example it happened with different "dye" stylists where you wish to change a lot of pallets and click 5-10x times then menu vanished by itself and you can't do anything.

Sounds like a npc timeout on npcs using a `menu` dialog without a `mes`-window?

**Comment by anacondaqq** [comment]
@Jeybla exactly!

**Comment by Lemongrass3110** [comment]
I just tried this and I am unable to reproduce this when I disable the SECURE_NPCTIMEOUT.
I can reproduce it when it is enabled though.

**Comment by anacondaqq** [comment]
@Lemongrass3110 no, this is not really related to SECURE_NPCTIMEOUT it was tested several times some time ago (maybe 1 month ago) and results the same with / without SECURE_NPCTIMEOUT, and yes, time to time it's looks like with disabled SECURE_NPCTIMEOUT the issue does not appears again, but it's not, just need faster to click enter (because looks like some delay inside)


### I have tested it a few minutes ago and...
![](https://i.imgur.com/SPBmj3x.png)

I still can reproduce the issue...

#### Let's talk about details:

1. talk script for reproducing g and load it to the game: https://pastebin.com/XuDyiFWx
2. now go to NPC and log-off (this is very important)
3. now login to the game and do not click anything, don't walk, etc, when you login first action what you need start to do it's left click on NPC at Prontera from the script above and press as fast as a possible enter button. Usually 5-10 times enough to reproduce it.
4. After around 5-10 times very fast pressing enter the menu will be removed and not showed up anymore, you will be still blocked by moving (because you have not unlocked moving at all).

**Important things**
I can't reproduce the issue if I log in and doing something in the game for a while.
I can reproduce issue always when login to the game and first what I do it's clicking on NPC and start to press enter button fast to switch menu

**Comment by Lemongrass3110** [comment]
Cant reproduce it with your script either. I even tried spamming the enter button on your script with MacroX.
It always waited for the dialog to show up before accepting the next enter...

**Comment by anacondaqq** [comment]
> It always waited for the dialog to show up before accepting the next enter...

Sorry, I'm on a laptop and can't write a video how to reproduce it. The problem not in "waiting for dialogue" the problem that something in client or emulator manually turns off showing the dialogue only when you are login to the game and do nothing, except clicking on NPC and pressing enter button 8-10 times fast, at the same time need to move cursor in screen (sounds crazy, I know), and need to try to click to any place outside the screen. 

If you can't to reproduce it, no problem at all, when I will drop some an overwhelming work, I will try to debug it and fix and share results.

**Comment by keitenai** [comment]
@Lemongrass3110 try adding this on a separated script:

`OnPCLoginEvent:
sleep2 1000;
end;
`

**Comment by Jeybla** [comment]
Thanks for your evaluation on this issue. From what I've seen by now this issue happens when you're in a NPC menu dialog and another script gets attached and ends. The attachment can be by sleep2 like @keitenai mentioned or just by using an addrid or attachrid.
I'd suggest you to avoid any recurring addrids/attachrids on your players and to minimize the `sleep2` interval in `OnPCLoginEvent`s until this issue is fixed.

**Comment by poporing** [comment]
@Jeybla this issue is happening with all npc from rathena having a switch menu. It's happen randomly and it can be on the first char you create and the first npc you talk.

I put back the info from the other issues that was closed:

On only a few players and on some NPC, a bug appears. the next disappears to leave a closed place.
if you click on close or it can autoclose too and then opens the menu ...
whatever he does, the player remains blocked. He can not do anything.
A message appears next if the player wants to talk to an NPC.
He is obliged to deco reco to regain his mobility and be able to speak to a NPC.


1> git version: 470364c
2> Client : 2015-11-04
3> NPC Offi from rathena (Ex: Captain Carroc > Izlude)
4> The bug appears on some players but not all, it's random.
5> It seems that the npcs that bug are those with a switch menu.
6> Version Re and Pre-re (Test on both)
7> No console error

for more information :

https://www.youtube.com/watch?v=OYebrfrLE6I

This is completly random and not happen to all players. It's happen on official script that have switch in the script from last git rathena, I'm using 2015-05-13 exe version.
I don't have any custom source and I didn't touch npc.
As you can see in the video when the script goes to the switch menu it's auto close then you are stuck
I have this problem since I start my server in october 2016
I try to disable SECURE_NPCTIMEOUT and completly remove it but I always got the problem.
There is no console error.

**Comment by rye305** [comment]
Can confirm this. Not all players have this issue. been sitting around 6 hours on this trying to fix it. Thought i fixed it but i didn't - it truly happens very randomly. Some players get stuck when interacting with the warper which is provided by rathena (unmodified)

**Comment by uddevil** [comment]
I think its not random, try this step : 
* player A click on warper, **do not choose anything**, leave the selection menu open
* player B use skill on player A, try different one from the list below : 
  * Standard Skill (just randomly casting skill using shift + click)
  * Damaging Skill (maybe using `@duel`)
  * Targeted Buff Skill (that change sc_)
  * Party Buff Skill (that change sc_ and must be in a party)
* one of those skills make the npc window vanishes, and player get stuck.
* somehow it happened with other npc that has long dialogues

**Comment by ecdarreola** [comment]
Just experienced this. And yeah for the warper it indeed freezes the character.

**Comment by RadianFord** [comment]
I encounter this issue before and all i can say is, its related to this part
```
/**
 * Optional NPC Dialog Timer
 * When enabled all npcs dialog will 'timeout' if user is on idle for longer than the amount of seconds allowed
 * - On 'timeout' the npc dialog window changes its next/menu to a 'close' button
 * Comment to disable the timer.
 **/
#define SECURE_NPCTIMEOUT
```
And try checking your scripts that runs a timer that is attached on the character.

**Comment by anacondaqq** [comment]
Can you please share what exe are you using and what diffs? Just print them here, we will compare.
I guess this is a client-side issue related to broken diff at NEMO, and not related to specific exe. Because I was able to reproduce the issue on 2012-04-10, and on 2015-11-04, and on 2013-08-07 and so on.

**Comment by uddevil** [comment]
@anacondaqq I'm using 2015-11-04 but i don't understand what you mean with "just print them here", i'm assuming that its from the log, here you go

```
6 Force Send Client Hash Packet
8 Custom Window Title
9 Disable 1rag1 type parameters (Recommended)
13 Disable Ragexe Filename Check (Recommended)
19 Enable Title Bar Menu
23 Enable /who command (Recommended)
24 Fix Camera Angles (Recommended)
28 Increase Headgear ViewID
29 Disable Game Guard (Recommended)
33 Always Call SelectKoreaClientInfo() (Recommended)
34 Enable /showname (Recommended)
36 Read msgstringtable.txt (Recommended)
37 Read questid2display.txt (Recommended)
38 Remove Gravity Ads (Recommended)
39 Remove Gravity Logo (Recommended)
40 Restore Login Window (Recommended)
41 Disable Nagle Algorithm (Recommended)
44 Translate Client (Recommended)
46 Use Normal Guild Brackets (Recommended)
47 Use Ragnarok Icon
48 Use Plain Text Descriptions (Recommended)
49 Enable Multiple GRFs (Recommended)
51 Always Show License Screen
52 Use Custom Font
53 Use Ascii on All LangTypes (Recommended)
64 @ Bug Fix (Recommended)
65 Load Custom lua file instead of iteminfo*.lub (Recommended)
71 Ignore Resource Errors
72 Ignore Missing Palette Error
73 Remove Hourly Announce (Recommended)
75 Enable Flag Emoticons
76 Enforce Official Login Background
81 Rename License File
84 Remove Serial Display (Recommended)
85 Show Cancel To Service Select
90 Enable DNS Support (Recommended)
91 Disconnect to Login Window
97 Cancel to Login Window (Recommended)
204 Increase Attack Display
207 Resize Font
213 Disable Help Message on Login (Recommended)
```

**Comment by Rolfxx** [comment]
@Jeybla any news on this issue? I can still reproduce it after checking all of the above
client: 20160203 
it is very redundant when you click on the default `npc/custom/warper.txt` , even tho this script doesnt attach the players.. 
it also happens in other npc, at locations without `mapflag loadevent`

**Comment by Jeybla** [comment]
@Rolfxx If you're able to reproduce the issue easily, please be free to share your test settings. Best if you'd describe the necessary steps to reproduce it with the current head revision. This could help to track the actual issue.

**Comment by reunite-ro** [comment]
Hello, Im also having this issue. yes, im using the hourly reward script that have 

OnPCLoginEvent:
attachnpctimer ""+strcharinfo(0)+"";﻿
initnpctimer;
end;

any way to solve this issue or better way to use for hourly reward script ?

**Comment by gidzdlcrz** [comment]
I can confirm that this bug still occur. Using 2015 and 2018 its still happening. Im using the default warper and the default quest shop by euphy. Weird. Is there a workaround?

**Comment by aleos89** [comment]
Can you guys try out #5260 and see if this is still happening?

**Comment by Latiosu** [comment]
> Can you guys try out #5260 and see if this is still happening?

This has resolved the issue for me on the NPCs where I experience it.

**Comment by bgamez23** [comment]
the bug still exist specially if you have script like OnPCLoginEvent. even you disabled the #define SECURE_NPCTIMEOUT it still close the menu option.

**Comment by aleos89** [comment]
Can you give us your script or an example of it?

**Comment by bgamez23** [comment]
`-	script	hourlypoints	-1,{

//--Start of the Script
OnPCLoginEvent:
attachnpctimer ""+strcharinfo(0)+"";
initnpctimer;
end;
    
OnTimer30000:
//Check if Vending (normal or @at)
if(checkvending() >= 1 || checkchatting() == 1) {
    dispbottom "The hourly points event stopped because you were vending / chatting. Please relog if you wish to start again.";
    stopnpctimer;
    end;
}

OnTimer60000:
set @minute, @minute + 1;
//Check for 1 Minute
if(@minute == 60){
    set @minute,0;
    getitem 512,1;
    dispbottom "You received 1 "+getitemname( 512 )+" by staying ingame for 1 hour";
    set @consecutive_hour, @consecutive_hour + 1;
    }
//Check for 12 hours consecutive
    if(@consecutive_hour == 6) {
    set @consecutive_hour,0;
    getitem 512,1;
    dispbottom "You received 6 "+getitemname( 512 )+" by staying ingame for consecutive of 6 hour";
    }
stopnpctimer;
initnpctimer;
end;

}`

**Comment by Daegaladh** [comment]
@bgamez23 your script doesn't have any kind of window dialog, how is that supposed to be the same bug?

**Comment by bgamez23** [comment]
> @bgamez23 your script doesn't have any kind of window dialog, how is that supposed to be the same bug?

the bug occurred once you had OnPCLoginEvent.even you disabled the settings of secure npc. the npc without dialog box still disappear.

**Comment by reunite-ro** [comment]
> @bgamez23 your script doesn't have any kind of window dialog, how is that supposed to be the same bug?

If I remember correctly, it happens when talking with other NPCs  when the script hourly enabled. It will happens randomly while talking to NPC such as Warper , Healer, etc.

**Comment by bgamez23** [comment]
> > @bgamez23 your script doesn't have any kind of window dialog, how is that supposed to be the same bug?
> 
> If I remember correctly, it happens when talking with other NPCs when the script hourly enabled. It will happens randomly while talking to NPC such as Warper , Healer, etc.

yes you are correct. as long as you have OnPCLoginEvent script on your npc you will get it randomly even you disabled the secure npc.

**Comment by Abszoluto** [comment]
> `- script hourlypoints -1,{
> 
> //--Start of the Script
> OnPCLoginEvent:
> attachnpctimer ""+strcharinfo(0)+"";
> initnpctimer;
> end;
> 
> OnTimer30000:
> //Check if Vending (normal or @at)
> if(checkvending() >= 1 || checkchatting() == 1) {
> dispbottom "The hourly points event stopped because you were vending / chatting. Please relog if you wish to start again.";
> stopnpctimer;
> end;
> }
> 
> OnTimer60000:
> set @minute, @minute + 1;
> //Check for 1 Minute
> if(@minute == 60){
> set @minute,0;
> getitem 512,1;
> dispbottom "You received 1 "+getitemname( 512 )+" by staying ingame for 1 hour";
> set @consecutive_hour, @consecutive_hour + 1;
> }
> //Check for 12 hours consecutive
> if(@consecutive_hour == 6) {
> set @consecutive_hour,0;
> getitem 512,1;
> dispbottom "You received 6 "+getitemname( 512 )+" by staying ingame for consecutive of 6 hour";
> }
> stopnpctimer;
> initnpctimer;
> end;
> 
> }`

I confirmed its on a different npc "OnPCLoginEvent:"
What should i do then ? there is anything to fix this ?
Already tried the hash, recompiled and still the same. The bug just stops when i disable the npc with "OnPCLoginEvent:"

**Comment by Lemongrass3110** [comment]
This issue should also be related to #5554

**Comment by Abszoluto** [comment]
> This issue should also be related to #5554

Just made this commit to my src and the problem still goes on..
Just noticed, the problem occours in every NPC in-game. I dont have any idea whats the problem. And my emulador is new, from 2020.

**Comment by gidzdlcrz** [comment]
The bug will happen if there is **attachnpctimer**

**Comment by Abszoluto** [comment]
> The bug will happen if there is **attachnpctimer** (Let's say hourly reward)

In any npc existing on the server ?

**Comment by Abszoluto** [comment]
Still having the issue... (already instaled #5554)

**Comment by Lemongrass3110** [comment]
I dont know why it was closed. I unlinked it before merging the PR.
Sorry.

**Comment by gravity-ro** [comment]
Confirm the problem is "initnpctimer" command if you remove it dialog work fine
i want to know how to fix it too.
